R/data.table.R

```diff
@@ -1195,7 +1195,7 @@ replace_dot_alias = function(e) {
           #   ok=-1 which will trigger setalloccol with verbose in the next
           #   branch, which again calls _selfrefok and returns the message then
           if ((ok<-selfrefok(x, verbose=FALSE))==0L)   # ok==0 so no warning when loaded from disk (-1) [-1 considered TRUE by R]
-            if (is.data.table(x)) warningf("Invalid .internal.selfref detected and fixed by taking a (shallow) copy of the data.table so that := can add this new column by reference. At an earlier point, this data.table has been copied by R (or was created manually using structure() or similar). Avoid names<- and attr<- which in R currently (and oddly) may copy the whole data.table. Use set* syntax instead to avoid copying: ?set, ?setnames and ?setattr. If this message doesn't help, please report your use case to the data.table issue tracker so the root cause can be fixed or this message improved.")
+            if (is.data.table(x)) warningf("A shallow copy of this data.table was taken so that := can add %d new columns by reference. At an earlier point, this data.table was copied by R (or was created manually using structure() or similar). Avoid names<- and attr<- which in R currently (and oddly) may copy the whole data.table. Use set* syntax instead to avoid copying: ?set, ?setnames and ?setattr. It's also common for data.table-unaware packages to produce affected tables. If this message doesn't help, please report your use case to the data.table issue tracker so the root cause can be fixed or this message improved.", length(newnames))
```

Comment:
Just found a new corner case when removing a non-existing column with invalid internal.selfref.
```r
DT = data.frame(x=1L, y=1:3)
setattr(DT, 'class', c("data.table", "data.frame"))
DT[, z := NULL]
```

Reply:
I think that's under

> or was created manually using structure() or similar

The wording looks OK to me, do you have a suggestion? Or maybe the trouble is it could look somewhat contradictory with the next sentence

> Use set* syntax instead to avoid copying

Where `setattr` was already used.

Reply:
I just find it strange that we speak of `add 1 new columns by reference` when we are in fact dropping a column.

Reply:
Interesting corner case indeed. At this point `newnames` is `setdiff(lhs, names(x))` where `lhs` is `"z"` and `names(x)` is (as expected) `c("x", "y")`. The code doesn't realise yet that the non-existent `z` is mentioned in the `j` expression only to be deleted.

Reply:
Simple fix for your case is` "add or remove", I guess it gets even messier when there are a mix of added+deleted columns in a query. I guess it will still be an improvement, though.